src/lib/Imp.hs

```diff
@@ -216,23 +216,19 @@ toImpOp (maybeDest, op) = case op of
     emitStatement (IDo, IThrowError)
     return $ Con $ AnyValue ty
   CastOp destTy x -> case (getType x, destTy) of
-    (BoolTy, PreludeBoolTy) -> do
-      tag <- emitInstr $ IPrimOp $ ScalarUnOp BoolToInt $ fromScalarAtom x
-      returnVal $ Con $ SumAsProd destTy (toScalarAtom IntTy tag) [[], []]
-    (PreludeBoolTy, BoolTy) -> case x of
-      DataCon _ [] 0 [] -> returnVal $ BoolLit $ Int8Lit 0
-      DataCon _ [] 1 [] -> returnVal $ BoolLit $ Int8Lit 1
-      Con (SumAsProd _ tag [[],[]]) -> do
-        ans <- emitInstr $ IPrimOp $ ScalarUnOp UnsafeIntToBool $ fromScalarAtom tag
-        returnVal $ toScalarAtom BoolTy ans
-      _ -> error $ "Not a prelude bool: " ++ pprint x
-    (BaseTy (Scalar Int64Type  ), IntTy ) -> castTo $ Scalar Int64Type
-    (BaseTy (Scalar Float64Type), FloatTy) -> castTo $ Scalar Float64Type
+    (_, BaseTy bt) -> castTo bt
+    (_, IntTy    ) -> castTo $ Scalar Int64Type
+    (_, FloatTy  ) -> castTo $ Scalar Float64Type
     _ -> error $ "Invalid cast: " ++ pprint (getType x) ++ " -> " ++ pprint destTy
     where
       castTo bt = do
-        result <- emitInstr (ICastOp (IValType bt) $ fromScalarAtom x)
+        result <- cast (fromScalarAtom x) bt
         returnVal $ toScalarAtom destTy result
+  Select p x y -> do
```

Comment:
Ah, answering my own question, I think. Do I understand right that this makes it possible to `Select` on arbitrary types?

Reply:
Yes, exactly! It's a little sub-optimal for those simple types right now, but as far as I can see we really only use `Select` to pick between the prelude bool values.